adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/MarkDuplicates.scala

```diff
@@ -17,153 +17,419 @@
+    fragmentsDf.groupBy('library, 'read1contigName, 'read1fivePrimePosition, 'read1strand) // left position
+      .agg(countDistinct('read2contigName, 'read2fivePrimePosition, 'read2strand) // right position
+        as 'groupCount)
   }
 
-  def apply(rdd: AlignmentRecordRDD): RDD[AlignmentRecord] = {
+  /**
+   * Marks each alignment record in the Dataset `alignmentRecords` as a duplicate based on duplicate information
+   * from a DataFrame specifying which fragments found in the collection of alignment records are duplicates.
+   * Each read will be marked as a duplicate if it is a primary alignments and part of a duplicate fragment
+   * or if it is a mapped read but not a primary alignment.
+   * Unmapped reads will not be marked as duplicate.
+   * @param alignmentRecords Dataset of AlignmentRecords
+   * @param duplicatesDf DataFrame containing information about each
+   * @return A DataFrame with the same schema as `alignmentRecords`  but with reads having
+   *         been marked as duplicates in the "duplicateRead" column in accordance with `duplicatesDf`
+   *         DataFrame which was provided.
+   */
+  private def markDuplicates(alignmentRecords: Dataset[AlignmentRecordSchema], duplicatesDf: DataFrame): DataFrame = {
+    import alignmentRecords.sparkSession.implicits._
+    addDuplicateFragmentInfo(alignmentRecords, duplicatesDf)
+      .withColumn("duplicateRead",
+        'duplicateFragment.isNotNull and 'readMapped and ('duplicateFragment or !'primaryAlignment))
+      .drop("duplicateFragment") // drop the temporary column for marking duplicate fragments
+  }
 
-    markBuckets(rdd.groupReadsByFragment(), rdd.recordGroups)
-      .flatMap(_.allReads)
+  /**
+   * Adds information about which fragments are duplicates given in `duplicatesDf` to a dataset of alignment records
+   * @param alignmentRecords Dataset of alignment records to add duplicate fragment info to
+   * @param duplicatesDf DataFrame with columns "recordGroupName", "readName" and "duplicateFragment" indicating
+   *                     for each fragment (identified by "record
+   * @return A new DataFrame identical to `alignmentRecords` but with an extra boolean column "duplicateFragment"
+   *         indicating for each alignment record whether or not it is part of a duplicate fragment
+   */
+  private def addDuplicateFragmentInfo(alignmentRecords: Dataset[AlignmentRecordSchema],
+                                       duplicatesDf: DataFrame): DataFrame = {
+    import alignmentRecords.sparkSession.implicits._
+    alignmentRecords.join(duplicatesDf, Seq("readName", "recordGroupName"), "left")
   }
 
-  def apply(rdd: FragmentRDD): RDD[Fragment] = {
+  /**
+   * Case class which merely extends the Fragment Schema by a single column "duplicateFragment" so that
+   * a DataFrame with fragments having been marked as duplicates can be cast back into a DataSet
+   */
+  private case class FragmentDuplicateSchema(readName: Option[String] = None,
```

Comment:
Do you think it would be useful to add `duplicateFragment` or similarly named field to the Avro schema definition for `Fragment`, or is this flag only useful in a temporary context?

Reply:
I'm not sure. I was considering it while developing but then I realized that it really only seems useful in a temporary context since each of the alignment records contained within the fragment also contain a duplicate flag so the schema has redundancies.

Reply:
I'm not sure. I was considering it while developing but then I realized that it really only seems useful in a temporary context since each of the alignment records contained within the fragment also contain a duplicate flag so the schema has redundancies.